### ISSUE TITLE
grains.remove does not remove nested dictionary

### ISSUE THREAD

**Issue by frogunder** [saltstack/salt#28414]
When I have a grain as follows: 

```
frode-test:
    ----------
    role:
        ----------
        tomcat:
            ----------
            instances:
                |_
                  ----------
                  key:
                      test2
```

When I try to remove this grain with the following command `salt \* grains.remove role:tomcat:instances "{ 'key': 'test2' }"` I get this output, it shows it is deleted but is is not:

```
frode-test:
    ----------
    role:tomcat:instances:
```

Here is my grains `salt \* cmd.run 'cat /etc/salt/grains'`
```
frode-test:
    role:
      tomcat:
        instances:
        - key: test2
    role:tomcat:instances: []
```

In fact it adds another grain `role:tomcat:instances: []`

Here is my salt version report:

```
           Salt: 2015.5.2
         Python: 2.7.5 (default, Jun 24 2015, 00:41:19)
         Jinja2: 2.7.3
       M2Crypto: 0.21.1
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.7.0
           RAET: Not Installed
            ZMQ: 4.0.5
           Mako: Not Installed
```

**Comment by frogunder** [comment]
ZD-#497

**Comment by basepi** [comment]
This has been fixed in 2015.8, and the fix will be in 2015.8.2.

Fix is here: https://github.com/saltstack/salt/pull/26945

Backport to 2015.8 is here: https://github.com/saltstack/salt/pull/28486
